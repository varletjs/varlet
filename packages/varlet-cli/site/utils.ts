import { get } from 'lodash-es'
import { StyleProvider, type Theme, withSiteConfigNamespace } from '@varlet/cli/client'

export interface Menu {
  doc: string
  text: Record<string, string>
  type: MenuTypes
  useMobile?: boolean
}

export enum MenuTypes {
  TITLE = 1,
  COMPONENT = 2,
  DOCUMENTATION = 3,
}

export function removeEmpty(object: Record<string, string> = {}) {
  return Object.keys(object).reduce((record: Record<string, string>, key) => {
    const value = object[key]
    value && (record[key] = value)
    return record
  }, {})
}

export function isPhone() {
  return /Android|webOS|iPhone|iPod|BlackBerry|Pad/i.test(navigator.userAgent)
}

export function inIframe() {
  return window.self !== window.top
}

export function setTheme(config: Record<string, any>, name: Theme) {
  const styleVars = withSiteConfigNamespace(get(config, name, {}))

  StyleProvider(styleVars)
}

export function utoa(data: string): string {
  return btoa(unescape(encodeURIComponent(data)))
}
