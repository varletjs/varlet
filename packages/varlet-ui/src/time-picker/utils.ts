import { hours24, hoursAmpm } from './props'
import type { AmPm, AllowedTime } from './props'
import { toNumber } from '@varlet/shared'

type DisableProps = {
  time: number
  format: string
  ampm: AmPm | undefined
  hour: string
  minute: number
  max: string | undefined
  min: string | undefined
  disableHour: Array<string>
  allowedTime: AllowedTime | undefined
}

export const notConvert = (format: string, ampm: AmPm | undefined): boolean => format === '24hr' || ampm === 'am'

export const convertHour = (format: string, ampm: AmPm | undefined, hour: string) => {
  const index = hoursAmpm.findIndex((hourAmpm) => toNumber(hourAmpm) === toNumber(hour))
  const getHour = notConvert(format, ampm) ? hour : hours24[index]

  return {
    hourStr: getHour,
    hourNum: toNumber(getHour),
  }
}

export const getNumberTime = (time: string) => {
  const [hour, minute, second] = time.split(':')

  return {
    hour: toNumber(hour),
    minute: toNumber(minute),
    second: toNumber(second),
  }
}

export const getIsDisableMinute = (values: Omit<DisableProps, 'minute'>): boolean => {
  const { time, format, ampm, hour, max, min, disableHour } = values
  const { hourStr, hourNum } = convertHour(format, ampm, hour)
  let isBetweenMinMax = false
  let isAllow = false

  if (disableHour.includes(hourStr)) return true

  if (max && !min) {
    const { hour: maxHour, minute: maxMinute } = getNumberTime(max)
    isBetweenMinMax = maxHour === hourNum && time > maxMinute
  }

  if (!max && min) {
    const { hour: minHour, minute: minMinute } = getNumberTime(min)

    isBetweenMinMax = minHour === hourNum && time < minMinute
  }

  if (max && min) {
    const { hour: maxHour, minute: maxMinute } = getNumberTime(max)
    const { hour: minHour, minute: minMinute } = getNumberTime(min)

    isBetweenMinMax = (minHour === hourNum && time < minMinute) || (maxHour === hourNum && time > maxMinute)
  }

  if (values.allowedTime?.minutes) isAllow = values.allowedTime?.minutes(time)

  return isBetweenMinMax || isAllow
}

export const getIsDisableSecond = (values: DisableProps): boolean => {
  const { time, format, ampm, hour, minute, max, min, disableHour } = values
  const { hourStr, hourNum } = convertHour(format, ampm, hour)
  let isBetweenMinMax = false
  let isAllow = false

  if (disableHour.includes(hourStr)) return true

  if (max && !min) {
    const { hour: maxHour, minute: maxMinute, second: maxSecond } = getNumberTime(max)

    isBetweenMinMax = (maxHour === hourNum && maxMinute < minute) || (maxMinute === minute && time > maxSecond)
  }

  if (!max && min) {
    const { hour: minHour, minute: minMinute, second: minSecond } = getNumberTime(min)

    isBetweenMinMax = (minHour === hourNum && minMinute > minute) || (minMinute === minute && time > minSecond)
  }

  if (max && min) {
    const { hour: maxHour, minute: maxMinute, second: maxSecond } = getNumberTime(max)
    const { hour: minHour, minute: minMinute, second: minSecond } = getNumberTime(min)

    isBetweenMinMax =
      (maxHour === hourNum && maxMinute < minute) ||
      (minHour === hourNum && minMinute > minute) ||
      (maxHour === hourNum && maxMinute === minute && time > maxSecond) ||
      (minHour === hourNum && minMinute === minute && time < minSecond)
  }

  if (values.allowedTime?.seconds) isAllow = values.allowedTime?.seconds(time)

  return isBetweenMinMax || isAllow
}
