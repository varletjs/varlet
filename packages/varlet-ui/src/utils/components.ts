import {
  createApp,
  h,
  getCurrentInstance,
  inject,
  onUnmounted,
  computed,
  provide,
  reactive,
  isVNode,
  onMounted,
  onBeforeUnmount,
  nextTick,
  ref,
  onActivated,
  onDeactivated,
} from 'vue'
import type { Component, VNode, ComputedRef, ComponentInternalInstance, Ref, ComponentPublicInstance } from 'vue'
import { isArray, removeItem } from '@varlet/shared'

export interface MountInstance {
  instance: ComponentPublicInstance
  unmount: () => void
}

export interface ChildrenCounter {
  collect(instance: ComponentInternalInstance): void

  clear(instance: ComponentInternalInstance): void

  instances: ComponentInternalInstance[]
}

export interface BaseParentProvider<C> {
  collect(childProvider: C): void

  clear(childProvider: C): void
}

export function pickProps<T, U extends keyof T>(props: T, propsKey: U): T[U]
export function pickProps<T, U extends keyof T>(props: T, propsKey: U[]): Pick<T, U>
export function pickProps(props: any, propsKey: any): any {
  return Array.isArray(propsKey)
    ? propsKey.reduce((pickedProps: any, key) => {
        pickedProps[key] = props[key]
        return pickedProps
      }, {})
    : props[propsKey]
}

export function mount(component: Component): MountInstance {
  const app = createApp(component)
  const host = document.createElement('div')
  document.body.appendChild(host)

  return {
    instance: app.mount(host),
    unmount() {
      app.unmount()
      document.body.removeChild(host)
    },
  }
}

export function mountInstance(
  component: any,
  props: Record<string, any> = {},
  eventListener: Record<string, any> = {}
): {
  unmountInstance: () => void
} {
  const Host = {
    setup() {
      return () =>
        h(component, {
          ...props,
          ...eventListener,
        })
    },
  }

  const { unmount } = mount(Host)
  return { unmountInstance: unmount }
}

export function flatVNodes(subTree: any) {
  const vNodes: VNode[] = []

  const flat = (subTree: any) => {
    if (subTree?.component) {
      flat(subTree?.component.subTree)
      return
    }

    if (Array.isArray(subTree?.children)) {
      subTree.children.forEach((child: any) => {
        if (isVNode(child)) {
          vNodes.push(child)

          flat(child)
        }
      })
    }
  }

  flat(subTree)

  return vNodes
}

export function useAtChildrenCounter(key: symbol) {
  const instances: ComponentInternalInstance[] = reactive([])
  const parentInstance: ComponentInternalInstance = getCurrentInstance() as ComponentInternalInstance

  const sortInstances = () => {
    const vNodes: any[] = flatVNodes(parentInstance.subTree)

    instances.sort((a, b) => {
      return vNodes.indexOf(a.vnode) - vNodes.indexOf(b.vnode)
    })
  }

  const collect = (instance: ComponentInternalInstance) => {
    instances.push(instance)
    sortInstances()
  }

  const clear = (instance: ComponentInternalInstance) => {
    removeItem(instances, instance)
  }

  provide<ChildrenCounter>(key, {
    collect,
    clear,
    instances,
  })

  const length: ComputedRef<number> = computed(() => instances.length)

  return {
    length,
  }
}

export function useAtParentIndex(key: symbol) {
  if (!keyInProvides(key)) {
    return { index: null }
  }

  const childrenCounter: ChildrenCounter = inject<ChildrenCounter>(key) as ChildrenCounter

  const { collect, clear, instances } = childrenCounter

  const instance: ComponentInternalInstance = getCurrentInstance() as ComponentInternalInstance

  onMounted(() => {
    nextTick().then(() => collect(instance))
  })
  onUnmounted(() => {
    nextTick().then(() => clear(instance))
  })

  const index = computed(() => instances.indexOf(instance))

  return {
    index,
  }
}

export function useChildren<P, C>(key: symbol) {
  const childProviders: C[] = []

  const collect = (childProvider: C) => {
    childProviders.push(childProvider)
  }

  const clear = (childProvider: C) => {
    removeItem(childProviders, childProvider)
  }

  const bindChildren = (parentProvider: P) => {
    provide<P & BaseParentProvider<C>>(key, {
      collect,
      clear,
      ...parentProvider,
    })
  }

  return {
    childProviders,
    bindChildren,
  }
}

export function useParent<P, C>(key: symbol) {
  if (!keyInProvides(key)) {
    return {
      parentProvider: null,
      bindParent: null,
    }
  }

  const rawParentProvider = inject<P & BaseParentProvider<C>>(key) as P & BaseParentProvider<C>

  const { collect, clear, ...parentProvider } = rawParentProvider

  const bindParent = (childProvider: C) => {
    onMounted(() => collect(childProvider))
    onBeforeUnmount(() => clear(childProvider))
  }

  return {
    parentProvider,
    bindParent,
  }
}

export function keyInProvides(key: symbol) {
  const instance = getCurrentInstance() as any

  return key in instance.provides
}

export function useValidation() {
  const errorMessage: Ref<string> = ref('')

  const validate = async (rules: any, value: any, apis?: any): Promise<boolean> => {
    if (!isArray(rules) || !rules.length) {
      return true
    }

    const resArr = await Promise.all(rules.map((rule) => rule(value, apis)))

    return !resArr.some((res) => {
      if (res !== true) {
        errorMessage.value = String(res)
        return true
      }

      return false
    })
  }

  const resetValidation = () => {
    errorMessage.value = ''
  }

  const validateWithTrigger = async <T>(validateTrigger: T[], trigger: T, rules: any, value: any, apis?: any) => {
    if (validateTrigger.includes(trigger)) {
      ;(await validate(rules, value, apis)) && (errorMessage.value = '')
    }
  }

  return {
    errorMessage,
    validate,
    resetValidation,
    validateWithTrigger,
  }
}

export function useRouteListener(cb: () => void) {
  onMounted(() => {
    window.addEventListener('hashchange', cb)
    window.addEventListener('popstate', cb)
  })
  onUnmounted(() => {
    window.removeEventListener('hashchange', cb)
    window.removeEventListener('popstate', cb)
  })
}

export function useTeleport() {
  const disabled: Ref<boolean> = ref(false)

  onActivated(() => {
    disabled.value = false
  })

  onDeactivated(() => {
    disabled.value = true
  })

  return {
    disabled,
  }
}

export function exposeApis<T = Record<string, any>>(apis: T) {
  const instance = getCurrentInstance()
  if (instance) {
    Object.assign(instance.proxy!, apis)
  }
}

type ClassName = string | undefined | null
type Classes = (ClassName | [any, ClassName, ClassName?])[]

export function createNamespace(name: string) {
  const namespace = `var-${name}`

  const createBEM = (suffix?: string): string => {
    if (!suffix) return namespace

    return suffix.startsWith('--') ? `${namespace}${suffix}` : `${namespace}__${suffix}`
  }

  const classes = (...classes: Classes): any[] => {
    return classes.map((className) => {
      if (isArray(className)) {
        const [condition, truthy, falsy = null] = className
        return condition ? truthy : falsy
      }

      return className
    })
  }

  return {
    n: createBEM,
    classes,
  }
}

export function call<F extends (...arg: any) => any, P extends Parameters<F>>(
  fn?: F | null,
  ...arg: P
): ReturnType<F> | undefined {
  if (fn) return fn(...arg)
}
