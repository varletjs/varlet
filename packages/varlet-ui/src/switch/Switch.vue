<template>
  <div class="var-switch">
    <div
      class="var-switch-block"
      :class="[disabled || formDisabled ? 'var-switch__disable' : null]"
      @click="switchActive"
      :style="styleComputed.switch"
    >
      <div
        :style="styleComputed.track"
        class="var-switch__track"
        :class="[
          modelValue === activeValue ? 'var-switch__track-active' : null,
          errorMessage ? 'var-switch__track-error' : null,
        ]"
      ></div>
      <div
        class="var-switch__ripple"
        :style="styleComputed.ripple"
        v-ripple="{
          disabled: !ripple || disabled || loading || formDisabled,
        }"
      >
        <div
          :style="styleComputed.handle"
          class="var-switch__handle var-elevation--2"
          :class="[
            modelValue === activeValue ? 'var-switch__handle-active' : null,
            errorMessage ? 'var-switch__handle-error' : null,
          ]"
        >
          <var-loading v-if="loading" :radius="toNumber(size) / 2 - 2" />
        </div>
      </div>
    </div>
    <var-form-details :error-message="errorMessage" />
  </div>
</template>

<script lang="ts">
import { defineComponent, computed, nextTick } from 'vue'
import { useValidation } from '../utils/components'
import { toNumber } from '../utils/shared'
import { useForm } from '../form/provide'
import { props } from './props'
import VarFormDetails from '../form-details'
import VarLoading from '../loading'
import Ripple from '../ripple'
import type { ComputedRef } from 'vue'
import type { SwitchProvider } from './provide'

type StyleProps = {
  width: string
  height: string
  left: string
  borderRadius: string
  backgroundColor: string
  color: string
}

export default defineComponent({
  name: 'VarSwitch',
  components: {
    VarLoading,
    VarFormDetails,
  },
  directives: { Ripple },
  props,
  setup(props) {
    const { bindForm, form } = useForm()
    const { errorMessage, validateWithTrigger: vt, validate: v, resetValidation } = useValidation()

    const validate = () => v(props.rules, props.modelValue)

    const validateWithTrigger = () => nextTick(() => vt(['onChange'], 'onChange', props.rules, props.modelValue))

    const styleComputed: ComputedRef<Record<string, Partial<StyleProps>>> = computed(() => {
      const { size, modelValue, color, closeColor, loadingColor, activeValue } = props

      const sizeNum = toNumber(size)
      const switchWidth = sizeNum * 2
      const switchHeight = sizeNum * 1.2

      return {
        handle: {
          width: `${size}px`,
          height: `${size}px`,
          backgroundColor: modelValue === activeValue ? color || '' : closeColor || '',
          color: loadingColor && loadingColor,
        },
        ripple: {
          left: modelValue === activeValue ? `${sizeNum / 2}px` : `-${sizeNum / 2}px`,
          color: modelValue === activeValue ? color || '' : closeColor || '#999',
          width: `${sizeNum * 2}px`,
          height: `${sizeNum * 2}px`,
        },
        track: {
          height: `${switchHeight * 0.6}px`,
          width: `${switchWidth - 2}px`,
          borderRadius: `${switchWidth / 3}px`,
          filter: modelValue === activeValue || errorMessage?.value ? 'opacity(.6)' : 'brightness(.6)',
          backgroundColor: modelValue === activeValue ? color || '' : closeColor || '',
        },
        switch: {
          height: `${switchHeight}px`,
          width: `${switchWidth}px`,
        },
      }
    })

    const switchActive = (event: MouseEvent) => {
      const {
        onClick,
        onChange,
        disabled,
        loading,
        readonly,
        modelValue,
        activeValue,
        inactiveValue,
        'onUpdate:modelValue': updateModelValue,
      } = props

      onClick?.(event)
      if (disabled || loading || readonly || form?.disabled.value || form?.readonly.value) return

      const newValue = modelValue === activeValue ? inactiveValue : activeValue
      onChange?.(newValue)
      updateModelValue?.(newValue)
      validateWithTrigger()
    }

    const reset = () => {
      props['onUpdate:modelValue']?.(props.inactiveValue)
      resetValidation()
    }

    const switchProvider: SwitchProvider = {
      reset,
      validate,
      resetValidation,
    }

    bindForm?.(switchProvider)

    return {
      switchActive,
      toNumber,
      styleComputed,
      errorMessage,
      formDisabled: form?.disabled,
      formReadonly: form?.readonly,
    }
  },
})
</script>

<style lang="less">
@import '../styles/common';
@import '../styles/elevation';
@import '../ripple/ripple';
@import '../loading/loading';
@import '../form-details/formDetails';
@import './switch';
</style>
