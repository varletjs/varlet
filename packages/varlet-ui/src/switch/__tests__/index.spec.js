import example from '../example'
import Switch from '..'
import VarSwitch from '../Switch'
import { mount } from '@vue/test-utils'
import { createApp } from 'vue'
import { delay } from '../../utils/jest'

test('test switch example', () => {
  const wrapper = mount(example)
  expect(wrapper.html()).toMatchSnapshot()
})

test('test switch plugin', () => {
  const app = createApp({}).use(Switch)
  expect(app.component(Switch.name)).toBeTruthy()
})

test('test switch value props and events', async () => {
  const clickFn = jest.fn()
  const changeFn = jest.fn()
  const template = `
    <var-switch v-model="value" @click="clickFn" @change="changeFn" :active-value="2" :inactive-value="1" />
  `

  const wrapper = mount({
    components: {
      [VarSwitch.name]: VarSwitch,
    },
    data() {
      return {
        value: 2,
      }
    },
    methods: {
      clickFn,
      changeFn,
    },
    template,
  })

  await delay(0)

  expect(wrapper.find('.var-switch__track').classes()).toContain('var-switch__track--active')

  await wrapper.find('.var-switch__block').trigger('click')

  expect(clickFn).toHaveBeenCalledTimes(1)
  expect(changeFn).toHaveBeenCalledTimes(1)

  await delay(0)

  expect(wrapper.vm.value).toBe(1)
  expect(wrapper.find('.var-switch__track').classes()).not.toContain('var-switch__track--active')
})

test('test switch not available', async () => {
  const clickFn = jest.fn()
  const changeFn = jest.fn()
  const template = `
    <var-switch v-model="value" @click="clickFn" @change="changeFn" :disabled="disabled" :readonly="readonly" />
  `

  const wrapper = mount({
    components: {
      [VarSwitch.name]: VarSwitch,
    },
    data() {
      return {
        value: true,
        disabled: true,
        readonly: false,
      }
    },
    methods: {
      clickFn,
      changeFn,
    },
    template,
  })

  await delay(0)

  expect(wrapper.find('.var-switch--disable')).toBeTruthy()

  await wrapper.find('.var-switch__block').trigger('click')

  expect(clickFn).toHaveBeenCalledTimes(1)
  expect(changeFn).toHaveBeenCalledTimes(0)
  expect(wrapper.vm.value).toBe(true)

  await wrapper.setData({
    disabled: false,
    readonly: true,
  })

  await wrapper.find('.var-switch__block').trigger('click')
  expect(clickFn).toHaveBeenCalledTimes(2)
  expect(changeFn).toHaveBeenCalledTimes(0)
  expect(wrapper.vm.value).toBe(true)
})

test('test switch color and size', async () => {
  const template = `
    <var-switch v-model="value" color="indigo" close-color="#89ddff" size="15" :ripple="false" />
  `
  const wrapper = mount({
    components: {
      [VarSwitch.name]: VarSwitch,
    },
    data() {
      return {
        value: true,
      }
    },
    template,
  })

  await delay(0)

  expect(wrapper.html()).toMatchSnapshot()

  await wrapper.setData({ value: false })

  await delay(0)

  expect(wrapper.html()).toMatchSnapshot()
})

test('test switch loading prop', () => {
  const wrapper = mount(VarSwitch, {
    props: {
      loading: true,
      modelValue: true,
      loadingColor: '#ff9800',
    },
  })

  expect(wrapper.html()).toMatchSnapshot()
})

test('test switch rules prop', async () => {
  const template = `<var-switch v-model="value" :rules="[(v) => v === true || '错误！']"/>`
  const wrapper = mount({
    components: {
      [VarSwitch.name]: VarSwitch,
    },
    data() {
      return {
        value: true,
      }
    },
    template,
  })

  await delay(0)

  expect(wrapper.html()).toMatchSnapshot()

  await wrapper.find('.var-switch__block').trigger('click')

  await delay(0)

  expect(wrapper.html()).toMatchSnapshot()
})
