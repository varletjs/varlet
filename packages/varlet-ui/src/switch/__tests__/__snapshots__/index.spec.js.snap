// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test switch color and size 1`] = `
"<div class=\\"var-switch\\">
  <div class=\\"var-switch__block\\" style=\\"height: 18px; width: 30px;\\">
    <div style=\\"height: 10.799999999999999px; width: 28.5px; border-radius: 10px; filter: opacity(.6); background-color: indigo;\\" class=\\"var-switch__track var-switch__track--active\\"></div>
    <div class=\\"var-switch__ripple var-switch__ripple--active\\" style=\\"left: 7.5px; color: indigo; width: 30px; height: 30px;\\">
      <div style=\\"width: 15px; height: 15px; background-color: indigo;\\" class=\\"var-switch__handle var-elevation--2 var-switch__handle--active\\">
        <!--v-if-->
      </div>
    </div>
  </div>
  <transition-stub>
    <!--v-if-->
  </transition-stub>
</div>"
`;

exports[`test switch color and size 2`] = `
"<div class=\\"var-switch\\">
  <div class=\\"var-switch__block\\" style=\\"height: 18px; width: 30px;\\">
    <div style=\\"height: 10.799999999999999px; width: 28.5px; border-radius: 10px; filter: brightness(.6); background-color: rgb(137, 221, 255);\\" class=\\"var-switch__track\\"></div>
    <div class=\\"var-switch__ripple\\" style=\\"left: -7.5px; color: rgb(137, 221, 255); width: 30px; height: 30px;\\">
      <div style=\\"width: 15px; height: 15px; background-color: rgb(137, 221, 255);\\" class=\\"var-switch__handle var-elevation--2\\">
        <!--v-if-->
      </div>
    </div>
  </div>
  <transition-stub>
    <!--v-if-->
  </transition-stub>
</div>"
`;

exports[`test switch example 1`] = `
"<div class=\\"app-type\\">基本使用</div>
<div class=\\"var-space var--box\\" style=\\"flex-direction: row; justify-content: start; flex-wrap: wrap; margin: -15.36px 0px;\\">
  <div style=\\"margin: 15.36px 0px;\\">
    <div class=\\"var-switch\\">
      <div class=\\"var-switch__block\\">
        <div style=\\"filter: opacity(.6);\\" class=\\"var-switch__track var-switch__track--active\\"></div>
        <div class=\\"var-switch__ripple var-switch__ripple--active\\">
          <div class=\\"var-switch__handle var-elevation--2 var-switch__handle--active\\">
            <!--v-if-->
          </div>
        </div>
      </div>
      <transition-stub>
        <!--v-if-->
      </transition-stub>
    </div>
  </div>
</div>
<div class=\\"app-type\\">不可用</div>
<div class=\\"var-space var--box\\" style=\\"flex-direction: row; justify-content: start; flex-wrap: wrap; margin: -15.36px 0px;\\">
  <div style=\\"margin: 15.36px 92.16px 15.36px 0px;\\">
    <div class=\\"var-switch\\">
      <div class=\\"var-switch__block var-switch--disable\\">
        <div style=\\"filter: opacity(.6);\\" class=\\"var-switch__track var-switch__track--active\\"></div>
        <div class=\\"var-switch__ripple var-switch__ripple--active\\">
          <div class=\\"var-switch__handle var-elevation--2 var-switch__handle--active\\">
            <!--v-if-->
          </div>
        </div>
      </div>
      <transition-stub>
        <!--v-if-->
      </transition-stub>
    </div>
  </div>
  <div style=\\"margin: 15.36px 0px;\\">
    <div class=\\"var-switch\\">
      <div class=\\"var-switch__block\\">
        <div style=\\"filter: opacity(.6);\\" class=\\"var-switch__track var-switch__track--active\\"></div>
        <div class=\\"var-switch__ripple var-switch__ripple--active\\">
          <div class=\\"var-switch__handle var-elevation--2 var-switch__handle--active\\">
            <!--v-if-->
          </div>
        </div>
      </div>
      <transition-stub>
        <!--v-if-->
      </transition-stub>
    </div>
  </div>
</div>
<div class=\\"app-type\\">自定义颜色</div>
<div class=\\"var-space var--box\\" style=\\"flex-direction: row; justify-content: start; flex-wrap: wrap; margin: -15.36px 0px;\\">
  <div style=\\"margin: 15.36px 92.16px 15.36px 0px;\\">
    <div class=\\"var-switch\\">
      <div class=\\"var-switch__block\\">
        <div style=\\"filter: opacity(.6);\\" class=\\"var-switch__track var-switch__track--active\\"></div>
        <div class=\\"var-switch__ripple var-switch__ripple--active\\">
          <div class=\\"var-switch__handle var-elevation--2 var-switch__handle--active\\">
            <!--v-if-->
          </div>
        </div>
      </div>
      <transition-stub>
        <!--v-if-->
      </transition-stub>
    </div>
  </div>
  <div style=\\"margin: 15.36px 0px;\\">
    <div class=\\"var-switch\\">
      <div class=\\"var-switch__block\\">
        <div style=\\"filter: opacity(.6); background-color: rgb(255, 159, 0);\\" class=\\"var-switch__track var-switch__track--active\\"></div>
        <div class=\\"var-switch__ripple var-switch__ripple--active\\" style=\\"color: rgb(255, 159, 0);\\">
          <div style=\\"background-color: rgb(255, 159, 0);\\" class=\\"var-switch__handle var-elevation--2 var-switch__handle--active\\">
            <!--v-if-->
          </div>
        </div>
      </div>
      <transition-stub>
        <!--v-if-->
      </transition-stub>
    </div>
  </div>
</div>
<div class=\\"app-type\\">不同大小</div>
<div class=\\"var-space var--box\\" style=\\"flex-direction: row; justify-content: start; align-items: center; flex-wrap: wrap; margin: -15.36px 0px;\\">
  <div style=\\"margin: 15.36px 112.64px 15.36px 0px;\\">
    <div class=\\"var-switch\\">
      <div class=\\"var-switch__block\\" style=\\"height: 5.1192vw; width: 8.532vw;\\">
        <div style=\\"height: 3.07152vw; width: 8.1054vw; border-radius: 2.844vw; filter: opacity(.6);\\" class=\\"var-switch__track var-switch__track--active\\"></div>
        <div class=\\"var-switch__ripple var-switch__ripple--active\\" style=\\"left: 2.133vw; width: 8.532vw; height: 8.532vw;\\">
          <div style=\\"width: 4.266vw; height: 4.266vw;\\" class=\\"var-switch__handle var-elevation--2 var-switch__handle--active\\">
            <!--v-if-->
          </div>
        </div>
      </div>
      <transition-stub>
        <!--v-if-->
      </transition-stub>
    </div>
  </div>
  <div style=\\"margin: 15.36px 112.64px 15.36px 0px;\\">
    <div class=\\"var-switch\\">
      <div class=\\"var-switch__block\\">
        <div style=\\"filter: opacity(.6);\\" class=\\"var-switch__track var-switch__track--active\\"></div>
        <div class=\\"var-switch__ripple var-switch__ripple--active\\">
          <div class=\\"var-switch__handle var-elevation--2 var-switch__handle--active\\">
            <!--v-if-->
          </div>
        </div>
      </div>
      <transition-stub>
        <!--v-if-->
      </transition-stub>
    </div>
  </div>
  <div style=\\"margin: 15.36px 0px;\\">
    <div class=\\"var-switch\\">
      <div class=\\"var-switch__block\\" style=\\"height: 7.68vw; width: 12.8vw;\\">
        <div style=\\"height: 4.608vw; width: 12.16vw; border-radius: 4.266666666666667vw; filter: opacity(.6);\\" class=\\"var-switch__track var-switch__track--active\\"></div>
        <div class=\\"var-switch__ripple var-switch__ripple--active\\" style=\\"left: 3.2vw; width: 12.8vw; height: 12.8vw;\\">
          <div style=\\"width: 6.4vw; height: 6.4vw;\\" class=\\"var-switch__handle var-elevation--2 var-switch__handle--active\\">
            <!--v-if-->
          </div>
        </div>
      </div>
      <transition-stub>
        <!--v-if-->
      </transition-stub>
    </div>
  </div>
</div>
<div class=\\"app-type\\">加载状态</div>
<div class=\\"var-space var--box\\" style=\\"flex-direction: row; justify-content: start; align-items: center; flex-wrap: wrap; margin: -15.36px 0px;\\">
  <div style=\\"margin: 15.36px 92.16px 15.36px 0px;\\">
    <div class=\\"var-switch\\">
      <div class=\\"var-switch__block\\">
        <div style=\\"filter: opacity(.6);\\" class=\\"var-switch__track var-switch__track--active\\"></div>
        <div class=\\"var-switch__ripple var-switch__ripple--active\\">
          <div class=\\"var-switch__handle var-elevation--2 var-switch__handle--active\\">
            <div class=\\"var-loading\\">
              <!--v-if-->
              <div class=\\"var--box var-loading__body\\">
                <div class=\\"var-loading__circle\\"><span class=\\"var-loading__circle-block var-loading__circle-block--normal\\" style=\\"width: 4.2664vw; height: 4.2664vw; color: currentColor;\\"><svg viewBox=\\"25 25 50 50\\"><circle cx=\\"50\\" cy=\\"50\\" r=\\"20\\" fill=\\"none\\"></circle></svg></span></div>
                <!--v-if-->
                <!--v-if-->
                <!--v-if-->
                <!--v-if-->
                <!--v-if-->
              </div>
            </div>
          </div>
        </div>
      </div>
      <transition-stub>
        <!--v-if-->
      </transition-stub>
    </div>
  </div>
  <div style=\\"margin: 15.36px 0px;\\">
    <div class=\\"var-switch\\">
      <div class=\\"var-switch__block\\" style=\\"height: 7.68vw; width: 12.8vw;\\">
        <div style=\\"height: 4.608vw; width: 12.16vw; border-radius: 4.266666666666667vw; filter: opacity(.6);\\" class=\\"var-switch__track var-switch__track--active\\"></div>
        <div class=\\"var-switch__ripple var-switch__ripple--active\\" style=\\"left: 3.2vw; width: 12.8vw; height: 12.8vw;\\">
          <div style=\\"width: 6.4vw; height: 6.4vw; color: rgb(255, 159, 0);\\" class=\\"var-switch__handle var-elevation--2 var-switch__handle--active\\">
            <div class=\\"var-loading\\">
              <!--v-if-->
              <div class=\\"var--box var-loading__body\\">
                <div class=\\"var-loading__circle\\"><span class=\\"var-loading__circle-block var-loading__circle-block--normal\\" style=\\"width: 5.120000000000001vw; height: 5.120000000000001vw; color: currentColor;\\"><svg viewBox=\\"25 25 50 50\\"><circle cx=\\"50\\" cy=\\"50\\" r=\\"20\\" fill=\\"none\\"></circle></svg></span></div>
                <!--v-if-->
                <!--v-if-->
                <!--v-if-->
                <!--v-if-->
                <!--v-if-->
              </div>
            </div>
          </div>
        </div>
      </div>
      <transition-stub>
        <!--v-if-->
      </transition-stub>
    </div>
  </div>
</div>
<div class=\\"app-type\\">值的校验</div>
<div class=\\"var-space var--box\\" style=\\"flex-direction: row; justify-content: start; flex-wrap: wrap; margin: -15.36px 0px;\\">
  <div style=\\"margin: 15.36px 0px;\\">
    <div class=\\"var-switch\\">
      <div class=\\"var-switch__block\\">
        <div style=\\"filter: opacity(.6);\\" class=\\"var-switch__track var-switch__track--active\\"></div>
        <div class=\\"var-switch__ripple var-switch__ripple--active\\">
          <div class=\\"var-switch__handle var-elevation--2 var-switch__handle--active\\">
            <!--v-if-->
          </div>
        </div>
      </div>
      <transition-stub>
        <!--v-if-->
      </transition-stub>
    </div>
  </div>
</div>"
`;

exports[`test switch loading prop 1`] = `
"<div class=\\"var-switch\\">
  <div class=\\"var-switch__block\\">
    <div style=\\"filter: opacity(.6);\\" class=\\"var-switch__track var-switch__track--active\\"></div>
    <div class=\\"var-switch__ripple var-switch__ripple--active\\">
      <div style=\\"color: rgb(255, 152, 0);\\" class=\\"var-switch__handle var-elevation--2 var-switch__handle--active\\">
        <div class=\\"var-loading\\">
          <!--v-if-->
          <div class=\\"var--box var-loading__body\\">
            <div class=\\"var-loading__circle\\"><span class=\\"var-loading__circle-block var-loading__circle-block--normal\\" style=\\"width: 4.2664vw; height: 4.2664vw; color: currentColor;\\"><svg viewBox=\\"25 25 50 50\\"><circle cx=\\"50\\" cy=\\"50\\" r=\\"20\\" fill=\\"none\\"></circle></svg></span></div>
            <!--v-if-->
            <!--v-if-->
            <!--v-if-->
            <!--v-if-->
            <!--v-if-->
          </div>
        </div>
      </div>
    </div>
  </div>
  <transition-stub>
    <!--v-if-->
  </transition-stub>
</div>"
`;

exports[`test switch rules prop 1`] = `
"<div class=\\"var-switch\\">
  <div class=\\"var-switch__block\\">
    <div style=\\"filter: opacity(.6);\\" class=\\"var-switch__track var-switch__track--active\\"></div>
    <div class=\\"var-switch__ripple var-switch__ripple--active\\">
      <div class=\\"var-switch__handle var-elevation--2 var-switch__handle--active\\">
        <!--v-if-->
      </div>
    </div>
  </div>
  <transition-stub>
    <!--v-if-->
  </transition-stub>
</div>"
`;

exports[`test switch rules prop 2`] = `
"<div class=\\"var-switch\\">
  <div class=\\"var-switch__block\\">
    <div style=\\"filter: opacity(.6);\\" class=\\"var-switch__track var-switch__track--error\\"></div>
    <div class=\\"var-switch__ripple\\" style=\\"color: rgb(153, 153, 153);\\">
      <div class=\\"var-switch__handle var-elevation--2 var-switch__handle--error\\">
        <!--v-if-->
      </div>
    </div>
  </div>
  <transition-stub>
    <div class=\\"var-form-details\\">
      <div class=\\"var-form-details__message\\">错误！</div>
      <div class=\\"var-form-details__length\\"></div>
    </div>
  </transition-stub>
</div>"
`;
